Implement IUOs in terms of Optionals that are force-unwrapped when needed.

rudkx opened the pull request:
This set of commits moves us from treating IUOs as a separate type in the type system, instead treating them as Optionals that are force-unwrapped as needed based on setting up disjunctions in the ConstraintSystem that attempt to first type-check these as Optional, and only if that fails attempts to type-check as the underlying type.

Commit:
IUO: Strip specifier type in IUO attribute verification.

Commit:
IUO: Add overload choice kind for IUO decls.

Add a new OverloadChoiceKind for decls that are either IUO-typed, or
function-typed meaning that the function result type is IUO-typed.

Not currently used, so NFC.

Commit:
Minor tweak: Make it clear an auto variable is a pointer.

Commit:
[IUO] Add a locator for the result of a cast to an IUO type.

Not used yet, so NFC.

Commit:
IUO: Add a method to build a binding disjunction for T? and T.

Not yet used, so NFC.

Commit:
IUO: Put the pieces in place to handle coercions and casts to IUO types.

For casts like 'x as T!' or 'x as! T!', generate a disjunction that
attempts to bind both the optional and non-optional T, preferring the
optional branch of the disjunction.

Note that this should effectively be NFC at the moment because we're
still generating the IUO type for T! rather than a plain optional, so
we first attempt the IUO type (as opposed to an Optional<T>) which
goes through existing logic that handles the IUOs as part of the type
system.

The new rewriting logic will actually do something once we switch over
to generating Optional<T> when T! is uttered.

Commit:
IUO: Look through SpecifierTypeRepr when adding IUO attributes.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
FYI, I'm splitting my changes out into bite-sized chunks and will likely be pushing more of them here before merging. This is currently a whole bunch of NFC because it doesn't actually switch over from treating 'T!' as 'T?'.

slavapestov commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -2331,6 +2331,44 @@ class ConstraintSystem {
+    underlyingType = type->getWithoutSpecifierType()->getAnyOptionalObjectType();
+    assert(underlyingType);
+
+    if (type->is<LValueType>())
```
Do both come up here? Maybe add comments explaining why/when

rudkx replied:
I suspect `InOut` shouldn't happen so I'll switch that to an assert in a separate commit.

Commit:
IUO: Set the IUO decl attribute on failable constructor decls.

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Verify that we do not expect to see SpecifierTypeRepr in typed patterns.

Commit:
IUO: Replace one ConstraintLocator with two.

Replace ImplicitlyUnwrappedCoercionResult with the more general
ImplicitlyUnwrappedValue.

Add ImplicitlyUnwrappedDisjunctionChoice as a marker that indicates
that we've already created a disjunction from an
ImplicitlyUnwrappedValue, in order to avoid infinite recursion during
binding.

Also add support to buildDisjunctionForImplicitlyUnwrappedOptional for
functions returning optionals.

NFC.

Commit:
IUO: Start creating the DeclForImplicitlyUnwrappedOptional OverloadChoice.

Use this in places where we have a decl that is marked with the
ImplicitlyUnwrappedOptionalAttr so that we can distinguish in the
solver which decls need to be potentially unwrapped in order to type
check successfully.

Commit:
Merge branch 'master' into iuo

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

Commit:
IUO: Create disjunctions for optional choices.

When binding an optional value, or function that returns an optional
value, if that value was produced from a decl that was declared an
IUO, create a disjunction.

This change appears to regress a couple diagnostics tests. I've opened
a bug report (https://bugs.swift.org/browse/SR-6626) to investigate
these so that we can produce the previous diagnostic again at some
point.

Commit:
IUO: Finish up support for subscripts and add a test.

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please test source compatibility

Commit:
IUO: Add ImplicitlyUnwrappedFunctionConversionExpr.

This expression node is only used when applying the results of
expression type checking. It initially appears above the function
reference that returns an optional that needs to be unwrapped, and
then when dealing with function application we remove this and insert
a node to force-unwrap the result of the function application.

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Make use of the disjunction choices in rewriting expressions.

After solving, make use of the disjunction choices in rewriting
expressions to force optionals where needed.

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Fix silly bug.

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Fix crasher regression.

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Do not short-circuit disjunctions if we had to apply fixes.

There are cases where the disjunctions created for IUOs can be solved
with the Optional choice if a fix is applied. We do not what to stop
with this solution as we may be able to solve the constraint system
without a fix by selecting the non-Optional side.

This restores a couple diagnostics that had changed in a previous
commit, but also slightly regresses another.

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Update the ClangImporter to surface whether decls are implicitly unwrapped.

Update several functions to return a Type/bool pair with the bool
representing whether the type is one that should be implicitly
unwrapped. This bool is surfaced up to the decl importing functions,
allowing them to set the appropriate decl attribute so that the
expression type checker will consider selecting the underlying type of
the optional involved during solving if the expression doesn't type
check with the optional type.

When importing Decls, use the IUO information to set the
ImplicitlyUnwrappedOptionalAttr attribute when appropriate.

We're still generating IUO types at the moment, so this change doesn't
really have much of an affect other than the creation of those
attributes.

rudkx commented:
@swift-ci Please test

rudkx commented:
Yay! (?) The swiftpm crash appears to reproduce on macOS now.

Commit:
IUO: Use getAnyOptionalObjectType() until we remove IUO from the type system.

Not doing so resulted in a failure to build swiftpm.

rudkx commented:
@swift-ci Please test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please test source compatibility

Commit:
IUO: Refactor some of the code that is forcing the results of operations.

In some cases we were missing forcing the result of newly created AST
nodes. This moves the forcing in the code that was creating these nodes.

Commit:
Merge branch 'master' into iuo

rudkx commented:
@swift-ci Please test

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Rework forceUnwrapResult to use a result type based on the expr argument.

rudkx commented:
@swift-ci Please test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please test source compatibility

Commit:
IUO: Fix an issue where we weren't forcing an IUO result in some cases.

rudkx commented:
@swift-ci Please test

Commit:
IUO: Update protocol conformance checking to check IUO attr on decl.

rudkx commented:
@swift-ci Please smoke test

Commit:
IUO: Updates to PrintAsObjC to use the decl attribute for IUO.

Use the attribute to update the OptionalTypeKind used for various
decls that we print.

rudkx commented:
@swift-ci Please test

Commit:
IUO: Rework failability verification and fix a verification failure.

rudkx commented:
@swift-ci Please test

Commit:
IUO: Fix the diagnostic for chaining from non-failable to failable init.

Rather than using the optional kind of the result type of the ctor we
chain to, check to see whether it has the IUO attribute set.

rudkx commented:
@swift-ci Please test

Commit:
IUO: Fixes for failability issues for imported ctors.

Commit:
IUO: Force cases where we go from optional to non-optional.

This used to look at the type and force only IUO cases and then hit an
unreachable on plain optionals.

We should never reach this point for plain optional declarations,
though, so we should just be able to force value conversion from
optional-to-non-optional, knowing that Sema only allowed those things
declared as implicitly unwrapped optionals though.

rudkx commented:
@swift-ci Please test

rudkx commented:
@DougGregor @xedin @slavapestov 

I am still working through failures and know there is at least another decl checking issue and some PrintAsObjC and IDE printing issues, a couple source compatibility issues, and probably other issues left to resolve, but I think this is in pretty good shape to take a first look at. I can resolve any feedback in future commits.

Commit:
Merge branch 'master' into iuo

Commit:
IUO: Fixup diagnoseMismatchedOptionals to look at IUO decl attribute.

rudkx commented:
I've been holding off merging this because of a source compatibility suite failure due to not having enough changes pushed out yet, but I don't want this to keep hanging, so I've started splitting out the individual changes and merging them separately.

rudkx commented:
I'm merging all of these as separate PRs so I no longer need this PR.
